### PR TITLE
State: Extract selectors to a global selectors file

### LINKED ISSUE #693: Consider adding Redux state selectors
As our state tree and the things we need to do with it become more complicated, we are starting to repeat ourselves a bit in the `connect` calls, and this will probably continue.  Also there is a good bit of testable logic starting to show up there. See #594 for a couple of examples.

To avoid repetition and make this code easier to test, we should consider adding something like [Calypso's state selectors](https://github.com/Automattic/wp-calypso/tree/master/client/state/selectors).

### PR DESCRIPTION
**Pull request by youknowriad**
closes #693 

To factorise some selectors and test them easily. I've moved all the redux selectors into a separate file and add unit tests.

### CODE REVIEW

File: editor/header/tools/publish-button.js [hunk 1]
```diff
@@ -61,15 +70,13 @@ function PublishButton( {
...
-		isRequesting: state.saving.requesting,
-		isSuccessful: state.saving.successful,
-		isError: !! state.saving.error,
-		requestIsNewPost: state.saving.isNew,
+		post: getCurrentPost( state ),
+		edits: getPostEdits( state ),
+		blocks: getBlocks( state ),
+		isRequesting: isSavingPost( state ),
+		isSuccessful: isPostSaveRequestSuccessful( state ),
+		isError: !! isPostSaveRequestFailed( state ),
```
Comment (aduth) [inline editor/header/tools/publish-button.js]: We can assume `isPostSaveRequestFailed` will return a boolean and drop the coercion

**Review by aduth — APPROVED** [review]
Selectors are a thing we consciously avoided by worry about boilerplate and indirection. I like that you've taken a mild approach with a single `selectors.js` file. Testing selectors seems very valuable. I'd be okay with this, but curious to hear if others have concerns.

**Comment by nylen** [comment]
> Selectors are a thing we consciously avoided by worry about boilerplate and indirection.

To address this, how about we make a rule (or rule of thumb) that we only extract something into a selector if it's more complicated than just a simple pull from state?  For example, this doesn't seem to add much value:

```js
export function isEditedPostDirty( state ) {
	return state.editor.dirty;
}
```

**Comment by aduth** [comment]
There is some value in treating state as an opaque shape with selectors as its isolated entry points; specifically that it leaves open the option to restructure the state object without requiring sweeping changes.

**Review by nylen — APPROVED** [review]
I don't have a clear preference either way.  Also, even if some of the tests are trivial now, this won't always be the case, and the additional testability will serve us well for any future refactoring of the state tree.

I'll rebase and merge this :+1:

**Pull request merged**